Jetty fails with IllegalStateException on hosts with many cores

On hosts with many cores (somewhere around 16 or more), Jetty throws an IllegalStateException when starting up. As the [webclient config parameters](https://github.com/eclipse/smarthome/blob/master/bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/internal/WebClientFactoryImpl.java#L61) are buried in the code, there's no obvious way to figure out how to work around the problem (other than finding [my post on the forum](https://community.openhab.org/t/fix-for-jetty-error-when-running-on-host-with-many-cores/57449)).

```
16:25:43.381 [ERROR] [et.http.internal.WebClientFactoryImpl] - Could not start Jetty http client
java.lang.IllegalStateException: Insufficient configured threads: required=13 < max=10 for QueuedThreadPool[common]@75f3f957{STARTED,5<=5<=10,i=5,q=0}[ReservedThreadExecutor@16f08868{s=0/1,p=0}]
        at org.eclipse.jetty.util.thread.ThreadPoolBudget.check(ThreadPoolBudget.java:149) ~[?:?]
        at org.eclipse.jetty.util.thread.ThreadPoolBudget.leaseTo(ThreadPoolBudget.java:130) ~[?:?]
        at org.eclipse.jetty.util.thread.ThreadPoolBudget.leaseFrom(ThreadPoolBudget.java:175) ~[?:?]
        at org.eclipse.jetty.io.SelectorManager.doStart(SelectorManager.java:251) ~[?:?]
        at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68) ~[?:?]
        at org.eclipse.jetty.util.component.ContainerLifeCycle.start(ContainerLifeCycle.java:138) ~[?:?]
        at org.eclipse.jetty.util.component.ContainerLifeCycle.doStart(ContainerLifeCycle.java:117) ~[?:?]
        at org.eclipse.jetty.client.AbstractConnectorHttpClientTransport.doStart(AbstractConnectorHttpClientTransport.java:64) ~[?:?]
        at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68) ~[?:?]
        at org.eclipse.jetty.util.component.ContainerLifeCycle.start(ContainerLifeCycle.java:138) ~[?:?]
        at org.eclipse.jetty.util.component.ContainerLifeCycle.doStart(ContainerLifeCycle.java:117) ~[?:?]
        at org.eclipse.jetty.client.HttpClient.doStart(HttpClient.java:241) ~[?:?]
        at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68) ~[?:?]
        at org.eclipse.smarthome.io.net.http.internal.WebClientFactoryImpl$2.run(WebClientFactoryImpl.java:258) ~[?:?]
        at org.eclipse.smarthome.io.net.http.internal.WebClientFactoryImpl$2.run(WebClientFactoryImpl.java:1) ~[?:?]
        at java.security.AccessController.doPrivileged(Native Method) ~[?:?]
        at org.eclipse.smarthome.io.net.http.internal.WebClientFactoryImpl.createHttpClientInternal(WebClientFactoryImpl.java:244) ~[?:?]
        at org.eclipse.smarthome.io.net.http.internal.WebClientFactoryImpl.access$7(WebClientFactoryImpl.java:242) ~[?:?]
        at org.eclipse.smarthome.io.net.http.internal.WebClientFactoryImpl$1.run(WebClientFactoryImpl.java:218) ~[?:?]
        at org.eclipse.smarthome.io.net.http.internal.WebClientFactoryImpl$1.run(WebClientFactoryImpl.java:1) ~[?:?]
        at java.security.AccessController.doPrivileged(Native Method) ~[?:?]
        at org.eclipse.smarthome.io.net.http.internal.WebClientFactoryImpl.initialize(WebClientFactoryImpl.java:209) ~[?:?]
        at org.eclipse.smarthome.io.net.http.internal.WebClientFactoryImpl.getCommonHttpClient(WebClientFactoryImpl.java:160) ~[?:?]
        at org.eclipse.smarthome.io.net.http.HttpUtil.executeUrlAndGetReponse(HttpUtil.java:194) ~[?:?]
        at org.eclipse.smarthome.io.net.http.HttpUtil.downloadData(HttpUtil.java:438) ~[?:?]
        at org.eclipse.smarthome.io.net.http.HttpUtil.downloadImage(HttpUtil.java:401) ~[?:?]
        at org.eclipse.smarthome.io.net.http.HttpUtil.downloadImage(HttpUtil.java:372) ~[?:?]
        at org.eclipse.smarthome.io.net.http.HttpUtil.downloadImage(HttpUtil.java:358) ~[?:?]
        at org.openhab.binding.squeezebox.internal.handler.SqueezeBoxPlayerHandler.lambda$0(SqueezeBoxPlayerHandler.java:419) ~[?:?]```

--- a/bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/internal/WebClientFactoryImpl.java
+++ b/bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/internal/WebClientFactoryImpl.java
@@ -129,14 +129,14 @@ public class WebClientFactoryImpl implements HttpClientFactory, WebSocketFactory
         Objects.requireNonNull(endpoint, "endpoint must not be null");
         logger.debug("http client for endpoint {} requested", endpoint);
         checkConsumerName(consumerName);
-        return createHttpClientInternal(consumerName, endpoint, false);
+        return createHttpClientInternal(consumerName, endpoint, false, null);
     }
 
     @Override
     public HttpClient createHttpClient(String consumerName) {
         logger.debug("http client for consumer {} requested", consumerName);
         checkConsumerName(consumerName);
-        return createHttpClientInternal(consumerName, null, false);
+        return createHttpClientInternal(consumerName, null, false, null);
     }
 
     @Override
@@ -145,14 +145,14 @@ public class WebClientFactoryImpl implements HttpClientFactory, WebSocketFactory
         Objects.requireNonNull(endpoint, "endpoint must not be null");
         logger.debug("web socket client for endpoint {} requested", endpoint);
         checkConsumerName(consumerName);
-        return createWebSocketClientInternal(consumerName, endpoint, false);
+        return createWebSocketClientInternal(consumerName, endpoint, false, null);
     }
 
     @Override
     public WebSocketClient createWebSocketClient(String consumerName) {
         logger.debug("web socket client for consumer {} requested", consumerName);
         checkConsumerName(consumerName);
-        return createWebSocketClientInternal(consumerName, null, false);
+        return createWebSocketClientInternal(consumerName, null, false, null);
     }
 
     @Override
@@ -215,12 +215,12 @@ public class WebClientFactoryImpl implements HttpClientFactory, WebSocketFactory
                         }
 
                         if (commonHttpClient == null) {
-                            commonHttpClient = createHttpClientInternal("common", null, true);
+                            commonHttpClient = createHttpClientInternal("common", null, true, threadPool);
                             logger.debug("Jetty shared http client created");
                         }
 
                         if (commonWebSocketClient == null) {
-                            commonWebSocketClient = createWebSocketClientInternal("common", null, true);
+                            commonWebSocketClient = createWebSocketClientInternal("common", null, true, threadPool);
                             logger.debug("Jetty shared web socket client created");
                         }
 
@@ -239,7 +239,8 @@ public class WebClientFactoryImpl implements HttpClientFactory, WebSocketFactory
         }
     }
 
-    private HttpClient createHttpClientInternal(String consumerName, @Nullable String endpoint, boolean startClient) {
+    private HttpClient createHttpClientInternal(String consumerName, @Nullable String endpoint, boolean startClient,
+            @Nullable QueuedThreadPool threadPool) {
         try {
             return AccessController.doPrivileged(new PrivilegedExceptionAction<HttpClient>() {
                 @Override
@@ -247,11 +248,15 @@ public class WebClientFactoryImpl implements HttpClientFactory, WebSocketFactory
                     logger.debug("creating http client for consumer {}, endpoint {}", consumerName, endpoint);
 
                     HttpClient httpClient = new HttpClient(createSslContextFactory(endpoint));
-                    final QueuedThreadPool queuedThreadPool = createThreadPool(consumerName, minThreadsCustom,
-                            maxThreadsCustom, keepAliveTimeoutCustom);
-
                     httpClient.setMaxConnectionsPerDestination(2);
-                    httpClient.setExecutor(queuedThreadPool);
+
+                    if (threadPool != null) {
+                        httpClient.setExecutor(threadPool);
+                    } else {
+                        final QueuedThreadPool queuedThreadPool = createThreadPool(consumerName, minThreadsCustom,
+                                maxThreadsCustom, keepAliveTimeoutCustom);
+                        httpClient.setExecutor(queuedThreadPool);
+                    }
 
                     if (startClient) {
                         try {
@@ -277,7 +282,7 @@ public class WebClientFactoryImpl implements HttpClientFactory, WebSocketFactory
     }
 
     private WebSocketClient createWebSocketClientInternal(String consumerName, @Nullable String endpoint,
-            boolean startClient) {
+            boolean startClient, @Nullable QueuedThreadPool threadPool) {
         try {
             return AccessController.doPrivileged(new PrivilegedExceptionAction<WebSocketClient>() {
                 @Override
@@ -285,10 +290,13 @@ public class WebClientFactoryImpl implements HttpClientFactory, WebSocketFactory
                     logger.debug("creating web socket client for consumer {}, endpoint {}", consumerName, endpoint);
 
                     WebSocketClient webSocketClient = new WebSocketClient(createSslContextFactory(endpoint));
-                    final QueuedThreadPool queuedThreadPool = createThreadPool(consumerName, minThreadsCustom,
-                            maxThreadsCustom, keepAliveTimeoutCustom);
-
-                    webSocketClient.setExecutor(queuedThreadPool);
+                    if (threadPool != null) {
+                        webSocketClient.setExecutor(threadPool);
+                    } else {
+                        final QueuedThreadPool queuedThreadPool = createThreadPool(consumerName, minThreadsCustom,
+                                maxThreadsCustom, keepAliveTimeoutCustom);
+                        webSocketClient.setExecutor(queuedThreadPool);
+                    }
 
                     if (startClient) {
                         try {

